generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -0,0 +1,312 @@
+
+abstract class BaseChipTop(val useIOCells: Boolean = true)(implicit val p: Parameters) extends RawModule {
+
+  val iocells = ArrayBuffer.empty[IOCell]
+
+  val systemClock = Wire(Input(Clock()))
+  val systemReset = Wire(Input(Bool()))
+
+  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+
+  def connectReset(r: Bool): Unit
+
+}
+
+trait HasChipTopSimpleClockAndReset { this: BaseChipTop =>
+  implicit val p: Parameters
+
+  val (clock, systemClockIO) = if (useIOCells) {
+    IOCell.generateIOFromSignal(systemClock, Some("iocell_clock"))
+  } else {
+    val port = IO(Input(Clock()))
+    systemClock := port
+    (port, Seq())
+  }
+
+  // TODO how to handle async vs sync reset?
+  val (reset, systemResetIO) = if (useIOCells) {
+    IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"))
+  } else {
+    val port = IO(Input(Bool()))
+    systemReset := port
+    (port, Seq())
+  }
+
+  iocells ++= systemClockIO
+  iocells ++= systemResetIO
+
+  def connectReset(r: Bool) { reset := r }
+
+}
+
+// TODO add a reset catch and sync block?
+
+trait CanHaveChipTopGPIO { this: BaseChipTop =>
+  implicit val p: Parameters
+
+  require(useIOCells, "Cannot use CanHaveChipTopGPIO without IO cells")
+
+  val gpio: Seq[Seq[Analog]] = system match {
+    case system: HasPeripheryGPIOModuleImp => {
+      system.gpio.map(_.pins).zipWithIndex.map { case (pins, i) =>
+        pins.zipWithIndex.map { case (pin, j) =>
+          val g = IO(Analog(1.W))
+          // TODO functionalize this?
+          val iocell = IOCell.exampleGPIO()
+          iocell.suggestName(s"iocell_gpio_${i}_${j}")
+          iocells += iocell
+          iocell.io.o := pin.o.oval
+          iocell.io.oe := pin.o.oe
+          iocell.io.ie := pin.o.ie
+          pin.i.ival := iocell.io.i
+          iocell.io.pad <> g
+          g
+        }
+      }
+    }
+    case _ => Seq()
+  }
+}
+
+trait CanHaveChipTopGPIOConnections { this: BaseChipTop =>
+
+  // TODO maybe this is overly restrictive?
+  require(!useIOCells, "Doesn't make sense to use CanHaveChipTopGPIOConnections with IO cells")
+
+  val gpio: Seq[GPIOPortIO] = system match {
+    case system: HasPeripheryGPIOModuleImp => {
+      system.gpio.map { g =>
+        val port = IO(new GPIOPortIO(g.c))
+        port <> g
+        port
+      }
+    }
+    case _ => Seq()
+  }
+
+}
+
+trait CanHaveChipTopUART { this: BaseChipTop =>
+  implicit val p: Parameters
+  val uart: Seq[UARTPortIO] = system match {
+    case system: HasPeripheryUARTModuleImp => {
+      system.uart.zipWithIndex.map { case (u, i) =>
+        if (useIOCells) {
+          val (port, ios) = IOCell.generateIOFromSignal(u, Some(s"iocell_uart_${i}"))
+          iocells ++= ios
+          port
+        } else {
+          val port = IO(new UARTPortIO())
+          port <> u
+          port
+        }
+      }
+    }
+    case _ => Seq()
+  }
+}
+
+trait CanHaveChipTopBlockDevice { this: BaseChipTop =>
```

Comment:
Do we need to plumb this stuff all the way out past ChipTop, given that this isn't the right way to tape-out a chip that supports blkdev?

In general, for stuff like NIC, blockdevice, masteraxi4, etc, is it fine to not provide plumbing between DigitalTop and ChipTop passing through those IOs, since those are sort of simulation-only devices?